crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -630,6 +630,81 @@ impl IndexerReader {
         stored_txn.try_into_iota_transaction_effects()
     }
 
+    /// Expensive check to assert whether all transactions
+    /// are indexed.
+    ///
+    /// It uses both the `tx_global_order` table
+    /// and the `checkpoints` table to cover old transactions.
+    fn deep_check_all_transactions_are_indexed(
+        &self,
+        digests: &[TransactionDigest],
+    ) -> Result<bool, IndexerError> {
+        let (checkpointed, optimistic) = self
+            .multi_get_transactions(digests)?
+            .into_iter()
+            .partition::<Vec<_>, _>(|tx| tx.checkpoint_sequence_number >= 0);
+        if !optimistic.is_empty() {
+            let num_optimistic = optimistic.len();
+            let optimistic_digests = optimistic
+                .into_iter()
+                .map(|tx| tx.transaction_digest)
+                .collect::<HashSet<_>>();
+            let num_indexed =
+                self.count_indexed_tx_global_orders(optimistic_digests.into_iter())?;
+            if num_indexed as usize != num_optimistic {
+                return Ok(false);
+            }
+        }
+        let Some(max_transaction_checkpoint) = checkpointed
+            .iter()
+            .map(|tx| tx.checkpoint_sequence_number)
+            .max()
+        else {
+            return Ok(false);
```

Comment:
Actually, if we have no checkpointed transactions to check, shouldn't we return true?